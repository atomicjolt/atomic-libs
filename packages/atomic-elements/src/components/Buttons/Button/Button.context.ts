import { ButtonProps } from ".";
import { createComponentContext } from "@utils/index";

export const ButtonContext = createComponentContext<ButtonProps>();
