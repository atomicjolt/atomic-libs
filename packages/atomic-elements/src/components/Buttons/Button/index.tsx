export { Button } from "./Button.component";
export type { ButtonProps, ButtonVariants } from "./Button.component";
