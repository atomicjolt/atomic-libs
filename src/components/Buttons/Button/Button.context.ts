import React from "react";
import { ButtonProps } from ".";
import { IconButtonProps } from "../IconButton";

export const ButtonContext = React.createContext<
  Partial<ButtonProps & IconButtonProps>
>({});
