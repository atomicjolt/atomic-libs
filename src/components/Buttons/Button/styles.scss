@use "../../defines" as *;
@use "../common.scss";

.aj-btn {
  &--primary {
    --btn-text-clr: var(--text-clr-inverted);
    --btn-bg-clr: var(--primary700);
    --btn-hover-text-clr: var(--btn-text-clr);
    --btn-hover-bg-clr: var(--primary800);
    @extend .aj-btn;
  }

  &--secondary {
    --btn-text-clr: var(--text-clr-alt);
    --btn-bg-clr: var(--neutral100);
    --btn-hover-text-clr: var(--text-clr);
    --btn-hover-bg-clr: var(--neutral200);
    --btn-border: var(--border);
    @extend .aj-btn;
  }

  &--error {
    --btn-text-clr: var(--text-clr-inverted);
    --btn-bg-clr: var(--error700);
    --btn-hover-text-clr: var(--btn-text-clr);
    --btn-hover-bg-clr: var(--error800);
    @extend .aj-btn;
  }

  &--success {
    --btn-text-clr: var(--text-clr-inverted);
    --btn-bg-clr: var(--success700);
    --btn-hover-text-clr: var(--btn-text-clr);
    --btn-hover-bg-clr: var(--success800);
    @extend .aj-btn;
  }

  &--inverted {
    --btn-text-clr: var(--text-clr);
    --btn-bg-clr: var(--neutral50);
    --btn-hover-text-clr: var(--btn-text-clr);
    --btn-hover-bg-clr: var(--btn-bg-clr);
    --btn-hover-shadow: 0 1px 3px hsla(221, 39%, 11%, 0.5);
    @extend .aj-btn;
  }
}
