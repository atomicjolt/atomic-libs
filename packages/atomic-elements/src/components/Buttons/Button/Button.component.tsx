import { forwardRef } from "react";
import { AriaButtonOptions } from "@react-aria/button";
import { mergeProps, useObjectRef } from "@react-aria/utils";

import { SpinnerLoader } from "../../Loaders/SpinnerLoader";
import {
  ExtendedSize,
  HasVariant,
  LoadingProps,
  RenderBaseProps,
  SuggestStrings,
} from "../../../types";
import { useFocusRing } from "@hooks/useFocusRing";
import { useRenderProps } from "@hooks/useRenderProps";
import { useButtonLink } from "@hooks/useButtonLink";
import { useContextProps } from "@hooks/useContextProps";
import { StyledButton } from "./Button.styles";
import { ButtonContext } from "./Button.context";
import { SlotProps } from "@hooks/useSlottedContext";

export type ButtonVariants = SuggestStrings<
  | "primary"
  | "secondary"
  | "link"
  | "success"
  | "error"
  | "inverted"
  | "content"
  | "border"
  | "ghost"
>;

interface ButtonRenderProps {
  isLoading: boolean;
  isPressed: boolean;
  isFocusVisible: boolean;
  isFocused: boolean;
}

export interface ButtonProps
  extends AriaButtonOptions<"button">,
    LoadingProps,
    RenderBaseProps<ButtonRenderProps>,
    HasVariant<ButtonVariants>,
    SlotProps {
  as?: "button" | "a";
  size?: ExtendedSize;
}

/** A button component that can be used to trigger actions or events
 *
 * @example <Button onPress={() => alert("Hello, world!")}>Click me</Button>
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, forwardedRef) {
    [props, forwardedRef] = useContextProps(ButtonContext, props, forwardedRef);

    const ref = useObjectRef(forwardedRef);

    const {
      isLoading = false,
      loadingLabel,
      loadingComplete = false,
      as = props.href ? "a" : "button",
      variant = "primary",
      size = "auto",
    } = props;

    const { buttonProps, isPressed } = useButtonLink(
      {
        ...props,
        elementType: as,
        "aria-label": isLoading ? loadingLabel : props["aria-label"],
      },
      ref
    );

    const { focusProps, isFocusVisible, isFocused } = useFocusRing();

    const renderProps = useRenderProps({
      componentClassName: "aje-btn",
      ...props,
      variant,
      size,
      values: { isPressed, isLoading, isFocusVisible, isFocused },
      selectors: {
        "data-pressed": isPressed,
        "data-loading": isLoading,
      },
    });

    return (
      <StyledButton
        as={as}
        ref={ref}
        {...mergeProps(buttonProps, focusProps, renderProps)}
      >
        {isLoading && (
          <SpinnerLoader
            isLoading={!loadingComplete}
            placement="absolute center"
          />
        )}
        {renderProps.children}
      </StyledButton>
    );
  }
);
