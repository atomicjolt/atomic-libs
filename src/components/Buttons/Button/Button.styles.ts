import styled from "styled-components";
import { BaseStyledButton } from "../Buttons.styles";
import mixins from "@styles/mixins";

export const StyledButton = styled(BaseStyledButton)`
  ${mixins.SizingX}
`;
