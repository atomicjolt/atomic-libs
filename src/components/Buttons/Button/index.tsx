import React from "react";
import '../../general.scss';
import "./styles.scss";
import Spinner from '../../Loaders/Spinner';
interface CommonProps {
  /** What to render within the Button */
  children?: React.ReactNode;
  className?: 'primary' | 'secondary' | 'success' | 'error' | 'inverted';
  type?: 'submit' | 'reset' | 'button';
  disabled?: boolean;
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
}

type LoadingProps =
  | { 
    loading?: false; 
    loadingLabel?: never;
    loadingComplete?: never;
  }
  | { 
    loading: true; 
    loadingLabel?: string;
    loadingComplete?: boolean;
  }

export type Props = CommonProps & LoadingProps;

/** Button Component */
export default function Button({ 
    children, 
    type = "button", 
    className, 
    disabled,
    loading,
    loadingLabel = 'loading',
    loadingComplete,
    onClick 
  }: Props) {

  let loadingClass = loading ? ' is-loading' : '';
  let loadingText = loading ? loadingLabel : '';
    
  return (
    <button
      aria-label={loadingText}
      type={type}
      className={`aj-btn--${className}${loadingClass}`}
      onClick={onClick}
      disabled={disabled}
    >
      {loading ?
        <Spinner loading={!loadingComplete} />
        : null
      }
      {children}
    </button>
  );
}
