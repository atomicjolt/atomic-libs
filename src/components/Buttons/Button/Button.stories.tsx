import { ComponentMeta, ComponentStory } from "@storybook/react";
import Button, { Props } from ".";

export default {
  title: "Buttons/Button",
  component: Button,
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args: Props) => (
  <Button {...args} />
);

export const Primary = Template.bind({});
Primary.args = {
  children: "Primary",
  className: "primary",
};

export const Secondary = Template.bind({});
Secondary.args = {
  children: "Secondary",
  className: "secondary",
};

export const Error = Template.bind({});
Error.args = {
  children: "Error",
  className: "error",
};

export const Success = Template.bind({});
Success.args = {
  children: "Success",
  className: "success",
};

export const Inverted = Template.bind({});
Inverted.args = {
  children: "Inverted",
  className: "inverted",
};
