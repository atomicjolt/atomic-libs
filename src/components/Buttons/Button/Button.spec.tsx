import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import Button from ".";

describe("matches snapshots", () => {
  test("normal display", () => {
    const result = render(<Button>Click Me</Button>);
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("changes className", () => {
    const result = render(<Button className="secondary">Click Me</Button>);
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("loading state", () => {
    const result2 = render(<Button loading={true}>Click Me</Button>);
    expect(result2.asFragment()).toMatchSnapshot();
  });

  test("loading completed state", () => {
    const result2 = render(
      <Button loading={true} loadingComplete={true}>
        Click Me
      </Button>
    );
    expect(result2.asFragment()).toMatchSnapshot();
  });
});

it("is clickable", () => {
  const onClick = jest.fn((e) => e);
  render(<Button onClick={onClick}>Click me</Button>);

  fireEvent.click(screen.getByText("Click me"));

  expect(onClick.mock.calls.length).toEqual(1);
});

it("displays loading displays spinner", () => {
  const { rerender } = render(
    <Button loading={true} loadingComplete={false}>
      Click Me
    </Button>
  );

  expect(screen.getByTestId("spinner-svg")).toBeInTheDocument();

  rerender(
    <Button loading={true} loadingComplete={true}>
      Click Me
    </Button>
  );

  expect(screen.queryByTestId("spinner-svg")).not.toBeInTheDocument();
});
