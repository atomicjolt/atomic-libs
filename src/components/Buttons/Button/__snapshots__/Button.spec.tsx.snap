// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshots > changes variant 1`] = `
<DocumentFragment>
  <button
    class="sc-kAycRU sc-imWZod cLCTCz gRIuSL aje-btn aje-btn--secondary is-auto"
    type="button"
  >
    Click Me
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshots > loading completed state 1`] = `
<DocumentFragment>
  <button
    class="sc-kAycRU sc-imWZod cLCTCz gRIuSL aje-btn aje-btn--primary is-auto"
    data-loading="true"
    type="button"
  >
    <div
      class="sc-gEvDqW foZKdl aje-loader aje-loader--spinner is-medium"
      data-placement="absolute center"
    >
      <span
        class="sc-fqkwJk eiXPrv aje-checkmark is-medium"
        data-testid="checkmark"
      />
    </div>
    Click Me
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshots > loading state 1`] = `
<DocumentFragment>
  <button
    class="sc-kAycRU sc-imWZod cLCTCz gRIuSL aje-btn aje-btn--primary is-auto"
    data-loading="true"
    type="button"
  >
    <div
      class="sc-gEvDqW foZKdl aje-loader aje-loader--spinner is-medium"
      data-placement="absolute center"
    >
      <div
        class="sc-aYaIB blILUf aje-flex"
      >
        <div
          class="sc-dcJtft cYhlUP aje-spinner is-medium"
        >
          <svg
            class="sc-iGgVNO HkWEM"
            data-testid="spinner-svg"
            viewBox="25 25 50 50"
          >
            <circle
              class="sc-gsFSjX jjboCs"
              cx="50"
              cy="50"
              fill="none"
              r="20"
              stroke-width="0.5em"
            />
          </svg>
        </div>
      </div>
    </div>
    Click Me
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshots > normal display 1`] = `
<DocumentFragment>
  <button
    class="sc-kAycRU sc-imWZod cLCTCz gRIuSL aje-btn aje-btn--primary is-auto"
    type="button"
  >
    Click Me
  </button>
</DocumentFragment>
`;
